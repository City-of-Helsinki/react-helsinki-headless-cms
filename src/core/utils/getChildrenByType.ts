import React from 'react';

/**
 * Gets the string type of the component or core html (JSX) element.
 * React Fragments will return type 'react.fragment'.
 * Priority will be given to the prop '__TYPE'.
 *
 * @param {ReactNode} component - The component to type check
 * @returns {string} - The string representation of the type
 */
export const typeOfComponent = (component) =>
  // eslint-disable-next-line no-underscore-dangle
  component?.props?.__TYPE ||
  component?.type
    ?.toString()
    .replace('Symbol(react.fragment)', 'react.fragment') ||
  undefined;

/**
 * Gets all children by specified type.
 * This function will check the prop '__TYPE' first
 * and then the 'type' string to match core html elements.
 * To find a React Fragment, search for type 'react.fragment'.
 *
 * @param {ReactNode} children - JSX children
 * @param {string[]} types - Types of children to match
 * @returns {ReactNode[]} - Array of matching children
 * @example
 * // Finds all occurrences of ToDo (custom component), div, and React Fragment
 * getChildrenByType(children, ['ToDo', 'div', 'react.fragment']);
 */
export const getChildrenByType = (children: React.ReactNode, types: string[]) =>
  React.Children.toArray(children).filter(
    (child) => types.indexOf(typeOfComponent(child)) !== -1,
  );
