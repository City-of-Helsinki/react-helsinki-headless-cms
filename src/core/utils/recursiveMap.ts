import React from 'react';

/**
 * Recursively loops through the children of a ReactNode
 * and runs the given function from arguments to the every child.
 *
 * If the child is a ReactElement and it has children,
 * the children will be recursively called with the given function.
 *
 * If the child is not a valid ReactElement,
 * which would mean that it is for example a string content of an HTML-element,
 * it is directly returned.
 *
 * See the unit tests for some examples.
 */
export const recursiveMap = (
  children: React.ReactNode,
  fn: (children: React.ReactNode) => React.ReactNode,
) =>
  React.Children.map(children, (child) => {
    if (!React.isValidElement(child)) {
      // Strings inside HTML elements, etc.
      return child;
    }

    if (child.props.children) {
      // Make a recursive call to find the last node
      return fn(
        React.cloneElement(child as React.ReactElement, {
          children: recursiveMap(child.props.children, fn),
        }),
      );
    }
    // Should never(?) come here
    return fn(child);
  })?.flat();
