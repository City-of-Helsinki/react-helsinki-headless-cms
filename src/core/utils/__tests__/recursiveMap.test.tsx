import React from 'react';

import { recursiveMap } from '../recursiveMap';
import { typeOfComponent } from '../getChildrenByType';

describe('recursiveMap', () => {
  const flattenedListItems = (
    <ul>
      <li>Should be included 1/2</li>
      <li>Should be included 2/2</li>
    </ul>
  );

  const nestedListItems = (
    <div>
      <ul>
        <li>Should be included 1/3</li>
      </ul>
      <ol>
        <li>Should be included 2/3</li>
        <li>Should be included 3/3</li>
      </ol>
    </div>
  );

  it.each([
    [flattenedListItems, jest.fn(), 2], // 2 x <li>
    [nestedListItems, jest.fn(), 1 + 1 + 3], // <ul>, <ol> and 3 x <li>
    [
      <div>
        <div>{nestedListItems}</div>
      </div>,
      jest.fn(),
      1 + 1 + 1 + 1 + 3, // <div>, sub <div>, <ul>, <ol> and 3 x <li>
    ],
  ])('recursively calls the defined function', (component, fn, callCount) => {
    const { children } = component.props;
    recursiveMap(children, fn);
    expect(fn).toHaveBeenCalledTimes(callCount);
  });

  it('can be used with void functions e.g. to count elements', () => {
    const { children } = nestedListItems.props;
    const elementCounts = {
      li: 3,
      ul: 1,
      ol: 1,
    };
    const elementCounter: Record<keyof typeof elementCounts, number> = {
      li: 0,
      ul: 0,
      ol: 0,
    };
    const fn = (child: React.ReactElement) => {
      switch (typeOfComponent(child)) {
        case 'li':
          elementCounter.li += 1;
          break;
        case 'ul':
          elementCounter.ul += 1;
          break;
        case 'ol':
          elementCounter.ol += 1;
          break;
        default:
          break;
      }
      return child;
    };
    recursiveMap(children, fn);
    expect(elementCounter).toStrictEqual(elementCounts);
  });

  it.each([
    [nestedListItems, 'li', 3],
    [
      <div>
        <p>1/3</p>
        <div>
          <p>2/3</p>
        </div>
        <div>
          <div>
            <p>3/3</p>
          </div>
        </div>
      </div>,
      'p',
      3,
    ],
  ])(
    'can be used with return values to collect content #',
    (component, type, elementCount) => {
      const { children } = component.props;
      const result = [];
      const fn = (child: React.ReactElement) => {
        if (typeOfComponent(child) === type) result.push(child);
        return child;
      };
      recursiveMap(children, fn);
      expect(result).toHaveLength(elementCount);
      expect(result).toMatchSnapshot();
    },
  );

  it('can be used to modify each children of a React node', () => {
    const {
      props: { children },
    } = (
      <>
        <div>A</div>
        <div>
          <div>B</div>
        </div>
        <div>
          <div>
            <div>C</div>
          </div>
        </div>
      </>
    );
    const fn = (child: React.ReactElement) =>
      React.cloneElement(child, { className: 'newTestClassInEveryElement' });
    const result = recursiveMap(children, fn);
    expect(result).toMatchSnapshot();
  });
});
