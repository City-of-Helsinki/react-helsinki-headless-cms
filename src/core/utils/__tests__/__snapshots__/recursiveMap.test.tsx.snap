// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`recursiveMap can be used to modify each children of a React node 1`] = `
[
  <div
    className="newTestClassInEveryElement"
  >
    A
  </div>,
  <div
    className="newTestClassInEveryElement"
  >
    <div
      className="newTestClassInEveryElement"
    >
      B
    </div>
  </div>,
  <div
    className="newTestClassInEveryElement"
  >
    <div
      className="newTestClassInEveryElement"
    >
      <div
        className="newTestClassInEveryElement"
      >
        C
      </div>
    </div>
  </div>,
]
`;

exports[`recursiveMap can be used with return values to collect content # 1`] = `
[
  <li>
    Should be included 1/3
  </li>,
  <li>
    Should be included 2/3
  </li>,
  <li>
    Should be included 3/3
  </li>,
]
`;

exports[`recursiveMap can be used with return values to collect content # 2`] = `
[
  <p>
    1/3
  </p>,
  <p>
    2/3
  </p>,
  <p>
    3/3
  </p>,
]
`;
